File: lib/shapes/line.ts
import { Shape } from "../ffi.ts";
import { LineOptions, LineReturn } from "../types.ts";
import { setLine, setRGB, setRGBA, setColor, setGradient } from "../utilty.ts";

export default function(data: Shape[], options?: LineOptions): LineReturn {
    if(!options) options ={};
    if(!options.x1) options.x1 = 10;
    if(!options.y1) options.y1 = 1;
    if(!options.x2) options.x2 = 20;
    if(!options.y2) options.y2 = 5;
    if(!options.color) options.color = {red: 0, green: 0, blue:0, alpha: 255};
    return { 
        x1:options.x1,
        y1:options.y1,
        x2:options.x2,
        y2:options.y2,
        color:options.color,
        draw(options?: LineOptions) { return draw(this, data, options) },
        setMove(x, y) {return setLine(this, x, y, 1)},
        setLine(x, y) {return setLine(this, x, y, 2)},
        setRGB(red, green, blue) { return setRGB(this, red, green, blue) as LineReturn },
        setRGBA(red, green, blue, alpha) { return setRGBA(this, red, green, blue, alpha) as LineReturn },
        setGradient(gradient) { return setGradient(this, gradient) as LineReturn },
        setColor(color) { return setColor(this, color) as LineReturn }
    }
}

function draw(line: LineReturn, data: Shape[], options?: LineOptions): LineReturn {
    if(!options) options ={};
    if(!options.x1) options.x1 = options.x1 == 0 ? 0 : options.x1 || line.x1;
    if(!options.y1) options.y1 = options.y1 == 0 ? 0 : options.y1 || line.y1;
    if(!options.x2) options.x2 = options.x2 == 0 ? 0 : options.x2 || line.x2;
    if(!options.y2) options.y2 = options.y2 == 0 ? 0 : options.y2 || line.y2;
    if(!options.color) options.color = options.color ? options.color : line.color;
    data.push({
        shape_type: "Line",
        data: [options.x1, options.y1, options.x2, options.y2],
        draw_type: "Stroke",
        text: null,
        color: options.color
    });
    return line
}

File: src/shapes/line.rs
use image::{Rgba, RgbaImage, load_from_memory};
use std::marker::PhantomData;
use raqote::{Source, SolidSource, GradientStop, Point, Gradient, Spread, PathBuilder, DrawTarget, LineCap, LineJoin, DrawOptions, StrokeStyle};
use super::super::{GradientSelect, Shape, Color};

pub fn draw_line_mut(img: &mut RgbaImage, shape: &Shape) {
    let (x1, y1, x2, y2) = (shape.data[0], shape.data[1], shape.data[2], shape.data[3]);

    let x: f32;
    let y: f32;
    if x1 > x2{
        x = x1;
    }
    else {
        x = x2;
    }
    if y1 > y2{
        y = y1;
    }
    else {
        y = y2;
    }
    let source: Source;
    match &shape.color {
        Color::RGBA { red, green, blue, alpha } => {
            source = Source::Solid(SolidSource{r: *red, g: *green, b: *blue, a: *alpha});
        }
        Color::Gradients { gradients } => {
            let mut stops:Vec<GradientStop> = Vec::new();
            for gradient in gradients.iter() {
                stops.push(GradientStop {position: gradient.position, color: raqote::Color::new(gradient.color[3], gradient.color[0], gradient.color[1], gradient.color[2])});
            }
            match &gradients[0].gradient_select {
                GradientSelect::Linear {start, end} => {
                    let start = Point { x: start.x, y: start.y, _unit: PhantomData };
                    let end = Point { x: end.x, y: end.y, _unit: PhantomData };
                    source = Source::new_linear_gradient(Gradient { stops }, start, end, Spread::Pad );
                }
                GradientSelect::Radial{center, radius} => {
                    let start = Point { x: center.x, y: center.y, _unit: PhantomData };
                    source = Source::new_radial_gradient(Gradient { stops }, start, *radius, Spread::Pad );
                }
            }
        }
    }
    let mut path_builder = PathBuilder::new();
    path_builder.move_to(x1, y1);
    path_builder.line_to(x2, y2);
    path_builder.close();
    let mut draw_target = DrawTarget::new(x as i32,y as i32);
    let style = StrokeStyle {
        width: 1.,
        cap: LineCap::Butt,
        join: LineJoin::Miter,
        miter_limit: 10.,
        dash_array: Vec::new(),
        dash_offset: 0.,
    };
    draw_target.stroke(&path_builder.finish(), &source, &style, &DrawOptions::new());
    let raw_image = draw_target.get_data_u8().to_vec();
    let buffer = load_from_memory(&raw_image).unwrap().to_rgba8();
    for (_x, _y, pixel,) in buffer.enumerate_pixels() {
        if pixel != &Rgba([0,0,0,0]) {
            img.put_pixel(_x, _y, pixel.to_owned());
        }
    }
}
